# BiG-MEx
BiG-MEx: a tool for the mining of Biosynthetic Gene Cluster (BGC) domains and classes in metagenomic data. It consists of the following modules:
1. run_bgc_dom_annot: fast identification of BGC protein domains.  
2. run_bgc_dom_div: BGC domain-based diversity analysis.  
3. run_bgc_class_pred: BGC class abundance predictions.  

## Citation
Pereira-Flores, E., Buttigieg, P. L., Medema, M. H., Meinicke, P., Gl√∂ckner, F. O. and Fernandez-Guerra, A.. (2018+). _Mining metagenomes for natural product biosynthetic gene clusters: unlocking new potential with ultrafast techniques_. Under review.

## Installation

BiG-MEx consists of four docker images: 
1. epereira/bgc_dom_annot  
2. epereira/bgc_dom_div  
3. epereira/bgc_dom_merge_div  
4. epereira/bgc_class_pred  

Before running BiG-MEx it is necessary to install [docker](https://www.docker.com/).

Then just clone the GitHub repository:
```
git clone git@github.com:pereiramemo/BiG-MEx.git
```

All four images are in [dockerhub](https://hub.docker.com/). These will be downloaded automatically the first time you run the scripts.

## Documentation

The run_bgc_\*.bash scripts run docker images, which include all the code, dependencies and data used in the analysis. Given that we are using [docker](https://www.docker.com/), if your user is not in the [docker group](https://docs.docker.com/engine/installation/linux/linux-postinstall/#manage-docker-as-a-non-root-user) in Linux or Mac OS, the run_bgc_\*.bash scripts have to be executed with sudo.

### 1. bgc_dom_annot
This first module runs [UProC](http://uproc.gobics.de/) using a BGC domain profile database. It takes as an input metagenomic unassembled data and outputs a BGC domain abundance profile table.

See help
```
./run_bgc_dom_annot.bash . . --help
```

### 2. bgc_dom_div

bgc_dom_div has two different modes: sample and merge. The sample mode takes as an input metagenomic unassembled data and generates a targeted assembly of the domains. Subsequently, it clusters the assembled sequences, places these on pre-computed reference trees and computes the Shannon diversity index. The merge mode integrates the sample mode results to compute the diversity based on rarefied subsamples and integrates the phylogenetic placement.  

See help
```
./run_bgc_dom_div.bash sample . . . --help

./run_bgc_dom_div.bash merge . . --help
```

### 3. bgc_class_pred
This module is based on the [bgcpred](https://github.com/pereiramemo/bgcpred) R package, which includes a library of BGC class abundance models. Based on the domain profile generated by bgc_dom_annot, this module computes the BGC class abundance profile.

See help
```
./run_bgc_class_pred.bash . . --help
```

### See the [wiki](https://github.com/pereiramemo/BiG-MEx/wiki) for further documentation.
