---
title: "Step-by-step bgcpred tutorial"
author: "Emiliano Pereira"
date: "October 20, 2017"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: show
---
# 1. Introduction

In the following tutorial we are going to create BGC class-specific models to predict the abundance of BGC classes in metagenomic samples. These models use the class abundance as the response variables, and the domains abundance of the corresponding classes, as the predictor variables. To construct a BGC class model, we apply a two-step zero-inflated process. First, we predict the presence or absence of the target BGC class using a random forest (RF) binary classifier (Breiman, 2001). Second, we use a multiple linear (ML) regression to predict the BGC class abundance, but only if the class was previously predicted as present. In the cases where the number of zero values is too low (<10) or non-existent, we directly apply a ML regression. The models are trained using a simulated metagenomic dataset of 150 samples (i.e. OMs dataset, see below).

# 2. Datasets

1. __OMs:__ 150 simulated metagenomes. It is based on the complete and nearly complete genome sequences of the Ocean Microbial Reference Gene Catalogue (OM-RGC) (Sunagawa et al., 2015). Composed of 376 different species.

2. __TGs:__ 150 simulated metagenomes. It is based on the complete and nearly complete genome sequences corresponding to the genera found in the TARA Oceans taxonomic annotation of the Operational Taxonomic Units (OTUs) (Sunagawa et al., 2015). Composed of 331 different species.

# 3. Step-by-step tutorial

## 3.1 Load libraries and tables
First we need to load the necessary libraries and data. Bgcpred installation instructions can be found [here](https://github.com/pereiramemo/bgcpred)

```{r, results='hide', message=FALSE, warning=FALSE }
library(tidyverse)
library(bgcpred)

OMs_domains <- read_csv(file = "data/150_domain_abund_simulated_OMs.csv", col_names = TRUE) %>%
  rename(metaG = X1)
TGs_domains <- read_csv(file = "data/150_domain_abund_simulated_TGs.csv", col_names = TRUE) %>%
  rename(metaG = X1)

OMs_classes <- read_csv(file = "data/150_class_abund_simulated_OMs.csv", col_names = TRUE) %>%
  rename(metaG = X1)
TGs_classes <- read_csv(file = "data/150_class_abund_simulated_TGs.csv", col_names = TRUE) %>%
  rename(metaG = X1)
```

## 3.2 Reformat TGs table

And massage a little bit the data. First we will add a few domains not present in the TGs_domains table but present in the OMs_domains table, this will make our life easier later for the prediction and training

```{r, warning=FALSE}
i <- !colnames(OMs_domains) %in% colnames(TGs_domains)
for (n in colnames(OMs_domains)[i]){
  TGs_domains <- TGs_domains %>%
    mutate_(.dots = setNames(0, n))
}
```

## 3.3 Create training and test data

Now is time to create our tables for the training and test of our models.

```{r, warning=FALSE}
data_train <- bind_rows(OMs_domains %>%
                          gather(bgc, counts, -metaG) %>%
                          group_by(metaG) %>%
                          mutate(rc = prop.table(counts)),
                          OMs_classes %>%
                          gather(bgc, counts, -metaG) %>%
                          group_by(metaG) %>%
                          mutate(rc = prop.table(counts))) %>%
  select(metaG, bgc, rc) %>%
  spread(bgc, rc, fill = 0) %>%
  ungroup()

data_test <- bind_rows(TGs_domains %>%
                         gather(bgc, counts, -metaG) %>%
                         group_by(metaG) %>%
                         mutate(rc = prop.table(counts)),
                       TGs_classes %>%
                         gather(bgc, counts, -metaG) %>%
                         group_by(metaG) %>%
                         mutate(rc = prop.table(counts))) %>%
  select(metaG, bgc, rc) %>%
  spread(bgc, rc, fill = 0) %>%
  ungroup()
```

## 3.4 Creating the models

For the model creation we will use the __get_domains()__ and __class_model_train()__ functions from bgcpred package (although the package already includes the models: see wrap_up_predict()).  

We will create the models for the lantipeptide, nrps, t1pks, t2pks, t3pks and terpene BGC classes. First, we use the get_domains() function to extract the corresponding domains of each BGC class. Then, we use these domains and classes RCs to train the models.

```{r, warning=FALSE}
bgc_classes <- c("lantipeptide","nrps","t1pks","t2pks","t3pks","terpene")

bgc_model <- list()

for (b in bgc_classes) {
  
  y <- data_train %>% select(b) %>% unlist
  domains_x <- get_domains(b)
  domains_x <- domains_x[ domains_x %in% colnames(data_train) ]
  x <- data_train %>% select(domains_x)
  bgc_model[[b]] <- class_model_train(y = y, 
                                      x = x, 
                                      binary_method = "rf", 
                                      regression_method = "lm", 
                                      seed = 111)
}
```

## 3.5 BGC class RCs predictions

Now that we have created the models, we will use them to predict the BGC class RCs in the TGs data set. For this task, we are going to use the __class_model_predict()__ function from bgcpred.

```{r, warning=FALSE}
bgc_pred <- list()

for (b in bgc_classes) {
  
  domains_x <- get_domains(b)
  domains_x <- domains_x[ domains_x %in% colnames(data_test) ]
  x <- data_test %>% select(domains_x)
  
  bgc_pred[[b]] <- class_model_predict(x = x, 
                                       model_c = bgc_model[[b]]$binary_model, 
                                       model_r = bgc_model[[b]]$regression_model)
}  
```

## 3.7 Results evaluation

Given that TGs is a simulated data set, we actually know the BGC class abundances, and we can compare those with the model predictions.

```{r, warning=FALSE}
# select some nice colors
class2color <- cbind(class = bgc_classes, 
                     color = c("#ECD078","#D95B43","#C02942",
                               "#542437","#53777A","#000000"))


# convert predictions list to a long data frame
pred_df <- bgc_pred %>%
  bind_cols() %>%
  gather(key = bgc, value = pred )


# extract and convert real abundances to a long data frame

ref_df <- test_data_classes %>% 
  select(bgc_classes) %>%
  gather(key = bgc, value = ref)


# join data frames
X <- bind_cols(pred_df, ref = ref_df$ref )


# make titles including correlation values
COR <- X %>% 
  group_by(bgc) %>%
  summarise(cor = cor(ref, pred) %>% round(., digits = 2))

titles <- paste(COR$bgc, "cor:", COR$cor)
names(titles) <- COR$bgc

# plot
ggplot(X, aes(x = ref, y=pred)) +
  geom_abline(intercept = 0, slope = 1, color = "gray60") +  
  geom_point(aes(color = bgc ), alpha = 0.5) +
  facet_wrap( ~ bgc, ncol = 2, nrow = 3, scales = "free", labeller = as_labeller(titles)) +
  scale_color_manual(values = class2color[,"color"], name = "BGC class") + 
  xlab("Reference abundance") +
  ylab("Predicted abundance") +
  expand_limits(y=0) +
  theme_light() +
  theme(strip.background = element_blank(),
        strip.text = element_text(color = "black", face = "bold")) +
  scale_x_continuous(labels = scales::percent) +
  scale_y_continuous(labels = scales::percent)
```


# 4. Checking MLR assumptions

## 4.1 Residuals normality  
Here we simply plot the distribution of the residuals.

```{r, warning=FALSE}
res_df <- lapply(bgc_model,"[[", "regression_model") %>%
          lapply(., residuals) %>%
          plyr::ldply(., cbind)

colnames(res_df) <- c("bgc","res")


# plot
ggplot(res_df, aes(res)) +
  geom_histogram(aes(fill = bgc), alpha = 0.8, bins = 30) +  
  facet_wrap( ~ bgc, ncol = 2, nrow = 2, scales = "free" ) +
  scale_fill_manual(values = class2color[,"color"], name = "BGC class") + 
  xlab("MLR residuals") +
  ylab("Counts") +
  expand_limits(y=0) +
  theme_light() +
  theme(strip.background = element_blank(),
        strip.text = element_text(color = "black", face = "bold"))



```

## 4.2 Homoscedasticity  
Now, we plot the residuals vs. the fitted values. The idea is to check if the variability of the residuals stays constant as the fitted values increase.

```{r, warning=FALSE}
fit_df <- lapply(bgc_model,"[[", "regression_model") %>%
                  lapply(., fitted) %>%
                  plyr::ldply(., cbind)

colnames(fit_df) <- c("bgc","fit")
fit2res_df <- data.frame(bgc = fit_df$bgc, fit = fit_df$fit, res = res_df$res)

# plot
ggplot(fit2res_df, aes(x = fit, y = res)) +
  geom_abline(intercept = 0, slope = 0, color = "gray60") +  
  geom_point(aes(color = bgc ), alpha = 0.5) + 
  facet_wrap( ~ bgc, ncol = 2, nrow = 2, scales = "free" ) +
  scale_color_manual(values = class2color[,"color"], name = "BGC class") + 
  xlab("Residuals") +
  ylab("Fitted") +
  expand_limits(y=0) +
  theme_light() +
  theme(strip.background = element_blank(),
        strip.text = element_text(color = "black", face = "bold"))



```

## 4.3 Linearity  
Lastly, we check if there is a linear relationship between the predictor and response variables. To do this, we will plot the residuals vs. the predictor variables. Given that we have to create a plot for each domain, we only show the plots for the nrps domains. 

```{r, warning=FALSE}
res_redu_df <- res_df %>% filter(bgc == "nrps")
domains_x <- get_domains("nrps")
domains_x <- domains_x[ domains_x %in% colnames(train_data_domains) ]

dom_abund2res_df <- data.frame(train_data_domains[ ,domains_x], res = res_redu_df$res) %>%
                    gather(key = "domain", value = "abund", domains_x )
  

# select domain colors
domain2colors <- c("#C02842","#B75131","#AE8539","#A0A540","#759D46")
        

ggplot(dom_abund2res_df, aes(x = abund, y = res)) +
  geom_abline(intercept = 0, slope = 0, color = "gray60") +  
  geom_point(aes(color = domain ), alpha = 0.5) + 
  facet_wrap( ~ domain, ncol = 2, nrow = 2, scales = "free" ) +
  scale_color_manual(values = domain2colors, name = "BGC domain") + 
  xlab("Abundance") +
  ylab("Residuals") +
  expand_limits(y=0) +
  theme_light() +
  theme(strip.background = element_blank(),
        strip.text = element_text(color = "black", face = "bold"))


```

# 5. Bibliography

* Breiman, L. (2001). Random forests. Machine Learning, 45(1), 5???32.
http://doi.org/10.1023/A:1010933404324  

* Delmont, T. O., Quince, C., Shaiber, A., Esen, O. C., Lee, S. T. M., Lucker, S., & Eren, A. M. (2017). Nitrogen-Fixing Populations Of Planctomycetes And  Proteobacteria Are Abundant In The Surface Ocean. bioRxiv, 129791. http://doi.org/10.1101/129791  

* Sunagawa, S., Coelho, L. P., Chaffron, S., Kultima, J. R., Labadie, K., Salazar, G., ??? Velayoudon, D. (2015). Structure and function of the global ocean microbiome. Science, 348(6237), 1261359???1261359. http://doi.org/10.1126/science.1261359   

# 6. Session Info
```{r, warning=FALSE}
sessionInfo()
```
